src/main/scala/firrtl/passes/MemUtils.scala

```diff
@@ -0,0 +1,174 @@
+      case (mt:BundleType, dt:BundleType) => seqCat((mt.fields zip dt.fields).map{ case (mf,df) =>
+        hiermask(WSubField(e, mf.name, mf.tpe, UNKNOWNGENDER), mf.tpe, df.tpe) })
+      case (mt:UIntType, dt) => groundBitMask(e, dt)
+      case (mt, dt) => error("Invalid type for mask component!")
+    }
+  }
+  def groundBitMask(e: Expression, dataType: Type): Expression = dataType match {
+      case dt:UIntType => seqCat(List.fill(bitWidth(dt).intValue)(e))
+      case dt:SIntType => seqCat(List.fill(bitWidth(dt).intValue)(e))
+      case dt => error("Mask type does not properly correspond with data type!")
+  }
+}
+
+object bitWidth {
+  def apply(dt: Type): BigInt = widthOf(dt)
+  def widthOf(dt: Type): BigInt = dt match {
+      case t:VectorType => t.size * bitWidth(t.tpe)
+      case t:BundleType => t.fields.map(f => bitWidth(f.tpe)).foldLeft(BigInt(0))(_+_)
+      case t:UIntType => baseWidth(t.width)
+      case t:SIntType => baseWidth(t.width)
+      case t => error("Unknown type encountered in bitWidth!")
+  }
+  def baseWidth(w: Width): BigInt = w match {
+      case IntWidth(width) => width
+      case w => error("Unknown width encountered in bitWidth!")
+  }
+}
+
+object fromBits {
+  def apply(lhs: Expression, rhs: Expression): Statement = {
+    val fbits = lhs match {
+      case ex: WRef => getPart(ex, ex.tpe, rhs, 0)
+      case ex: WSubField => getPart(ex, ex.tpe, rhs, 0)
+      case ex: WSubIndex => getPart(ex, ex.tpe, rhs, 0)
+      case t => error("Invalid LHS expression for fromBits!")
+    }
+    Block(fbits._2)
+  }
+  def getPartGround(lhs: Expression, lhst: Type, rhs: Expression, offset: BigInt): (BigInt, Seq[Statement]) = {
+    val intWidth = bitWidth(lhst)
+    val sel = DoPrim(PrimOps.Bits, Seq(rhs), Seq(offset+intWidth-1,offset), UnknownType)
+    (offset + intWidth, Seq(Connect(NoInfo,lhs,sel)))
+  }
+  def getPart(lhs: Expression, lhst: Type, rhs: Expression, offset: BigInt): (BigInt, Seq[Statement]) = {
+    lhst match {
+      case t:VectorType => {
+        var currentOffset = offset
+        var stmts = Seq.empty[Statement]
+        for (i <- (0 to t.size)) {
+          val (tmpOffset, substmts) = getPart(WSubIndex(lhs, i, t.tpe, UNKNOWNGENDER), t.tpe, rhs, currentOffset)
+          stmts = stmts ++ substmts
+          currentOffset = tmpOffset
+        }
+        (currentOffset, stmts)
+      }
+      case t:BundleType => {
+        var currentOffset = offset
+        var stmts = Seq.empty[Statement]
+        for (f <- t.fields.reverse) {
+          val (tmpOffset, substmts) = getPart(WSubField(lhs, f.name, f.tpe, UNKNOWNGENDER), f.tpe, rhs, currentOffset)
+          stmts = stmts ++ substmts
+          currentOffset = tmpOffset
+        }
+        (currentOffset, stmts)
+      }
+      case t:UIntType => getPartGround(lhs, t, rhs, offset)
+      case t:SIntType => getPartGround(lhs, t, rhs, offset)
```

Comment:
Yeah seems like it's the same number of lines and less clear to get the same strictness.